### PR TITLE
Scheduler refactoring

### PR DESCRIPTION
**Pull request by gliargovas**

### CODE REVIEW

File: parallel-orch/orch.py [hunk 1]
```diff
@@ -52,10 +52,10 @@ def cmd_execution_info_simplified(cmd_execution_info):
     for cmd in cmd_execution_info.values():
         cmd.log_simplified()
 
-def log_run_and_workset_info(partial_program_order, reps, workset):
+def log_run_and_workset_info(partial_program_order, reps):
     logging.debug(f"=" * 20)
     logging.debug(f"RUN:{reps}")
-    logging.debug(f"WORKSET:{[str(partial_program_order.get_node(node_id)) for node_id in workset]}")
+    logging.debug(f"WORKSET:{[str(partial_program_order.get_node(node_id)) for node_id in partial_program_order.get_workset()]}")
```
Comment (angelhof) [inline parallel-orch/orch.py]: This can move to be a method of partial program order

File: parallel-orch/orch.py [hunk 2]
```diff
@@ -1,26 +1,34 @@
...
 import logging
 import executor
 import trace
 import util
 
+## TODO: For much later, ignore for now. We can discover W-R dependencies and stream their outputs
+##       instead of just waiting for execution to complete.
+
 # TODO: Currently cmd_execution_info does not create correct r/w sets for
 #       commands with same first part but different redir.
```
Comment (angelhof) [inline parallel-orch/orch.py]: Remove this obsolete comment (we now deal with IDs)!

File: parallel-orch/orch.py [hunk 3]
```diff
@@ -32,87 +40,29 @@ def parse_args(args=sys.argv[1:]):
...
-    def get_suffix(self, i):
-        return self.list_of_cmds[i+1:]
-
-## cmd_execution_info is a dictionary containing information about each command.
-## id : Cmd_exec_info (id, command, read set, write set, is commited)
-def generate_cmd_execution_info(cmds_to_run):
-    cmds_exec_info_format = [Cmd_exec_info(cmd) for cmd in cmds_to_run]
-    return {cmd.id: cmd for cmd in cmds_exec_info_format}
 
 ## cmd_to_id is a dictionary that maps full commands to their ids.
```
Comment (angelhof) [inline parallel-orch/orch.py]: Is this function relevant anymore?

File: parallel-orch/orch.py [hunk 4]
```diff
@@ -122,28 +72,22 @@ def generate_cmd_to_id(cmd_exec_info):
...
-    read_set, write_set = trace.parse_and_gather_cmd_rw_sets(cmd, trace_object)
-    
-    # Update the sets in cmd_execution_info
-    cmd_execution_info[cmd].update_read_set(read_set)
-    cmd_execution_info[cmd].update_write_set(write_set)
-    return cmd_execution_info
+    read_set, write_set = trace.parse_and_gather_cmd_rw_sets(trace_object)
+    return RWSet(read_set, write_set)
 
 ## FIXME: Read sets are not generated correctly for nested reads.
```
Comment (angelhof) [inline parallel-orch/orch.py]: Is this comment relevant? I think that now we can simply add every read/write in the trace to the RWset right?

File: parallel-orch/orch.py [hunk 5]
```diff
@@ -166,141 +110,64 @@ def add_launch_assignments_to_rw_sets(cmd_execution_info, trace_object):
                     if rhs in open_refs[command_prefix]:
                         path_ref = open_refs[command_prefix][rhs]
                         if trace.is_path_ref_read(path_ref):
-                            cmd_execution_info[launch_name].add_to_read_set(trace.get_path_ref_name(path_ref))
+                            # cmd_execution_info[launch_name].add_to_read_set(trace.get_path_ref_name(path_ref))
```
Comment (angelhof) [inline parallel-orch/orch.py]: Delete this if obsolete

File: parallel-orch/orch.py [hunk 6]
```diff
@@ -166,141 +110,64 @@ def add_launch_assignments_to_rw_sets(cmd_execution_info, trace_object):
...
-            elif second_cmd_id not in new_workset and \
-               has_backward_dependency(cmd_execution_info, first_cmd_id, second_cmd_id):
-                new_workset.insert_at_end(second_cmd_id)
-                logging.debug(f"Backward dependency: {first_cmd_id}, {second_cmd_id}")
-            elif second_cmd_id not in new_workset and \
-               has_write_dependency(cmd_execution_info, first_cmd_id, second_cmd_id):
-                new_workset.insert_at_end(second_cmd_id)
-                logging.debug(f"Write dependency: {first_cmd_id}, {second_cmd_id}")
-    return new_workset
+                            # cmd_execution_info[launch_name].add_to_write_set(trace.get_path_ref_name(path_ref))
```
Comment (angelhof) [inline parallel-orch/orch.py]: ditto

File: parallel-orch/orch.py [hunk 7]
```diff
@@ -166,141 +110,64 @@ def add_launch_assignments_to_rw_sets(cmd_execution_info, trace_object):
...
+                            node_id = partial_program_order.get_node_id_from_cmd_no_redir(launch_name)
+                            partial_program_order.add_to_write_set(node_id, trace.get_path_ref_name(path_ref))
 
 def workset_cmds_to_list(cmd_execution_info):
```
Comment (angelhof) [inline parallel-orch/orch.py]: delete(?)

File: parallel-orch/orch.py [hunk 8]
```diff
@@ -166,141 +110,64 @@ def add_launch_assignments_to_rw_sets(cmd_execution_info, trace_object):
...
         cmds_to_run.append(cmd.cmd)
     return cmds_to_run
 
-## Warning! HACK: Get rid of these functions on a later iteration.
-## TODO: We should maybe use a more efficient way 
-##       to pass the cmd_exec_info structure to trace      
-def convert_cmd_exec_info_to_cmd_based_dict(cmd_execution_info):
-    return {trace.remove_command_redir(cmd_obj.cmd): cmd_obj for cmd_obj in cmd_execution_info.values()}
-
 def convert_cmd_exec_info_cmd_based_to_id_based_dict(cmd_execution_info_cmd_based_key):
```
Comment (angelhof) [inline parallel-orch/orch.py]: delete?

File: parallel-orch/orch.py [hunk 9]
```diff
@@ -166,141 +110,64 @@ def add_launch_assignments_to_rw_sets(cmd_execution_info, trace_object):
...
     return {cmd_obj.id: cmd_obj for cmd_obj in cmd_execution_info_cmd_based_key.values()}
 
-def execute_workset_and_find_rw_dependencies(cmd_execution_info, workset):
+def execute_workset_and_find_rw_dependencies(partial_program_order: PartialProgramOrder):
     ## Warning! HACK: Remove these functions in later iteration
```
Comment (angelhof) [inline parallel-orch/orch.py]: delete comment?

File: parallel-orch/orch.py [hunk 10]
```diff
@@ -318,41 +185,30 @@ def run_and_trace_workset(workset, cmd_execution_info):
...
-    # cmd_to_id = generate_cmd_to_id(cmd_execution_info)
-
+def scheduling_algorithm(partial_program_order):
     # The workset contains all the command ids that are going to be traced in the current cycle
-    workset = Workset([cmd.id for cmd in cmd_execution_info.values()])
+    partial_program_order.init_workset()
     # Count tracing cycles
     reps = 1
-    ## Parse trace
     ## TODO: This will change when we actually hook up with riker
```
Comment (angelhof) [inline parallel-orch/orch.py]: This comment is obsolete I think

File: parallel-orch/partial_program_order.py [hunk 11]
```diff
@@ -0,0 +1,240 @@
...
+        return sorted(list(self.speculated))
+
+    def init_inverse_adjacency(self):
+        self.inverse_adjacency = {i: [] for i in self.nodes.keys()}
+        for from_id, to_ids in self.adjacency.items():
+            for to_id in to_ids:
+                self.inverse_adjacency[to_id].append(from_id)
+
+    # ## TODO: (When there is time) Define a function that checks that the graph is valid
+    def valid(self):
```
Comment (angelhof) [inline parallel-orch/partial_program_order.py:89]: Try to assert this function at the start and end of big functions (like resolve_dependencies) (when in debug mode)

File: parallel-orch/template_script_to_execute_in_overlay.sh [hunk 12]
```diff
@@ -5,6 +5,8 @@ echo $CMD_STRING > ./Rikerfile
 
 ## Call Riker to execute the command
 # rkr --debug --log all --show
+echo "------"
```
Comment (angelhof) [inline parallel-orch/template_script_to_execute_in_overlay.sh]: Are these echoes useful? Maybe we can just make them happen when in debug mode? Or just delete them for now

**Review by angelhof — COMMENTED** [review]
Looks great! Congrats :) Could you resolve my comments and then we can merge!

File: parallel-orch/orch.py (hunk 4 above)
Comment (gliargovas) [reply parallel-orch/orch.py]: Yes, this solution also works for nested R/W sets

**Pull request merged**